File: Benchmarking/noscaling/euler35/euler35.cpp
//
//  main.cpp
//  Euler35
//
//  Created by Sonni on 21/05/16.
//  Copyright © 2016 Sonni Jensen. All rights reserved.
//

#include <iostream>
#include <vector>

std::vector<bool> getBoolPrimes(unsigned int size)
{
    std::vector<bool> array;
    for (unsigned int i = 0; i <= size; i++)
        array.push_back(true);
    
    for (unsigned int i = 2; i <= size; i++)
    {
        if (array[i] == true)
        {
            unsigned int j = i + i;
            while (j <= size)
            {
                array[j] = false;
                j += i;
            }
        }
    }
    
    return array;
}

std::vector<unsigned int> getPrimes(std::vector<bool> array, unsigned int size)
{
    std::vector<unsigned int> primeArray;
    for (unsigned int i = 2; i <= size; i++)
        if (array[i])
            primeArray.push_back(i);
    
    return primeArray;
}

std::string rotate(std::string s)
{
    std::string first = s.substr(0, 1);
    std::string last = s.substr(1, s.length());
    return last + first;
}

unsigned int calc(unsigned int limit)
{
    unsigned int count = 0;
    std::vector<bool> primesBool = getBoolPrimes(limit);
    std::vector<unsigned int> primes = getPrimes(primesBool, limit);
    
    for (unsigned int i = 0; i < primes.size(); i++)
    {
        std::string s = std::to_string(primes[i]);
        unsigned int l = s.size();
        bool circPrime = true;
        for (unsigned int i = 0; i < l; i++)
        {
            s = rotate(s);
            int n = std::stoi(s);
            if (!primesBool[n])
            {
                circPrime = false;
                break;
            }
        }
        if (circPrime)
            count++;
    }
    return count;
}

int main(int argc, const char * argv[]) {
    calc(atoi(argv[1]));
    return 0;
}

File: Benchmarking/noscaling/euler35/euler35.py
import sys

limit = int(sys.argv[1])

def getPrimes(size):
	array = [True for i in xrange(size + 1)]
	array[1] = False
	c = 1
	for i in xrange(2, size + 1):
		if array[i] == True:
			j = i + i
			while j <= size:
				array[j] = False
				j += i
			c += 1
	primeArray = []
	for i in xrange(2, size + 1):
		if(array[i]):
			primeArray.append(i)
	return primeArray, array

def rotate(s):
  first = s[0]
  last = s[1:]
  return last + first

def calc(limit):
  primes, primesBool = getPrimes(limit)
  count = 0
  for p in primes:
    s = str(p)
    l = len(s)
    circPrime = True
    for i in xrange(1, l):
      s = rotate(s)
      n = int(s)
      if not(primesBool[n]):
        circPrime = False
        break
    if circPrime:
      count += 1
  return count
  
calc(limit)
